File: backend/app/schemas/annotation_batch.py
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class AnnotationBatchCreate(BaseModel):
    dataset_id: uuid.UUID
    name: str = Field(min_length=1, max_length=255)
    description: str | None = None
    image_ids: list[uuid.UUID] = Field(min_length=1)
    assignee_user_id: uuid.UUID | None = None


class AnnotationBatchAction(BaseModel):
    """Reserved for future action metadata; actions intentionally accept an empty body."""


class AnnotationBatchItemResponse(BaseModel):
    id: uuid.UUID
    batch_id: uuid.UUID
    image_id: uuid.UUID
    annotator_user_id: uuid.UUID | None
    status: str
    image_filename: str
    image_width: int
    image_height: int
    image_split: str
    created_at: datetime
    updated_at: datetime


class AnnotationBatchResponse(BaseModel):
    id: uuid.UUID
    dataset_id: uuid.UUID
    name: str
    description: str | None
    status: str
    assignee_user_id: uuid.UUID | None
    created_by_user_id: uuid.UUID | None
    total_count: int
    completed_count: int
    dataset_name: str | None = None
    assignee_name: str | None = None
    created_by_name: str | None = None
    created_at: datetime
    updated_at: datetime
    items: list[AnnotationBatchItemResponse] = Field(default_factory=list)


class AnnotationReviewAction(BaseModel):
    comment: str | None = Field(default=None, max_length=2000)
    quality_score: float | None = Field(default=None, ge=0, le=100)


class AnnotationReviewResponse(BaseModel):
    id: uuid.UUID
    batch_item_id: uuid.UUID
    image_id: uuid.UUID
    annotator_user_id: uuid.UUID | None
    reviewer_user_id: uuid.UUID | None
    status: str
    quality_score: float | None
    comment: str | None
    reviewed_at: datetime | None
    created_at: datetime
    updated_at: datetime
    batch_id: uuid.UUID
    batch_name: str | None = None
    dataset_id: uuid.UUID | None = None
    dataset_name: str | None = None
    image_filename: str | None = None
    annotation_count: int = 0
    item_status: str | None = None
    annotator_name: str | None = None
    reviewer_name: str | None = None

File: frontend/src/types/annotation-batch.ts
export type AnnotationBatchStatus =
  | 'draft'
  | 'assigned'
  | 'in_progress'
  | 'submitted'
  | 'completed'
  | 'cancelled'

export type AnnotationBatchItemStatus =
  | 'pending'
  | 'annotating'
  | 'submitted'
  | 'approved'
  | 'rejected'

export interface AnnotationBatchItem {
  id: string
  batch_id: string
  image_id: string
  annotator_user_id: string | null
  status: AnnotationBatchItemStatus | string
  image_filename: string
  image_width: number
  image_height: number
  image_split: string
  created_at: string
  updated_at: string
}

export interface AnnotationBatch {
  id: string
  dataset_id: string
  name: string
  description: string | null
  status: AnnotationBatchStatus | string
  assignee_user_id: string | null
  created_by_user_id: string | null
  total_count: number
  completed_count: number
  dataset_name: string | null
  assignee_name: string | null
  created_by_name: string | null
  created_at: string
  updated_at: string
  items: AnnotationBatchItem[]
}

export interface AnnotationReview {
  id: string
  batch_item_id: string
  image_id: string
  annotator_user_id: string | null
  reviewer_user_id: string | null
  status: 'pending' | 'approved' | 'rejected' | string
  quality_score: number | null
  comment: string | null
  reviewed_at: string | null
  created_at: string
  updated_at: string
  batch_id: string
  batch_name: string | null
  dataset_id: string | null
  dataset_name: string | null
  image_filename: string | null
  annotation_count: number
  item_status: string | null
  annotator_name: string | null
  reviewer_name: string | null
}
